README.md

ICON CLI tool
========

 ICON supports Command Line Interface(CLI interface) for 3rd party or user services development. With this single tool, you can control all ICON functions and automate them using scripts.

<!-- TOC depthFrom:1 depthTo:3 withLinks:1 updateOnSave:1 orderedList:0 -->

- [Prerequisite](#prerequisite)
- [Version](#version)
- [Glossary](#glossary)
- [Technical information](#technical-information)
- [Getting started](#getting-started)
	- [Installation](#installation)
	- [Run CLI](#run-cli)
	- [Console instructions](#console-instructions)
	- [Wallet operation](#wallet-operation)
		- [Create wallet file](#create-wallet-file)
		- [Show wallet information](#show-wallet-information)
		- [List up all assets in current wallet](#list-up-all-assets-in-current-wallet)
		- [Transfer the value to the specific address with the fee.](#transfer-the-value-to-the-specific-address-with-the-fee)

<!-- /TOC -->


# Prerequisite

You need Python 3.6 or later to run icxcli. You can have multiple Python versions (2.x and 3.x) installed on the same system without problems.

On macOS(Sierra and after) and Ubuntu 16.04, for example, the necessary packages are:

- automake
- pkg-config
- virtualenv
- autoconf
- libtool

In Ubuntu 16.04, you can install Python 3.6.

    $ sudo add-apt-repository ppa:jonathonf/python-3.6
    $ sudo apt update
    $ sudo apt install python3.6
    $ sudo apt install python3.6-dev
    $ sudo apt install virtualenv
    $ virtualenv -p python3.6 .venv

You can install pip3 for python 3.6 like this:

    $ wget https://bootstrap.pypa.io/get-pip.py
    $ sudo python3.6 get-pip.py

You can install autoconf, libtool and automake for python3.6.

    $ sudo apt install autoconf libtool automake



# Version

- 0.0.4



# Glossary

- Address of wallet: Unique string to identify the address to transfer value. It prefixed with 'hx'.
- Private key: A tiny bit of code that is paired with a public key to set off algorithms to encrypt and decrypt a text for the specific address.
- Public key: Long alphanumeric characters that is used to encrypt data (message).



# Technical information

## Private key, public key and wallet address

There are five steps to get from  private->public -> address:

1. Generate a private key.
2. Derive a public key from the private key.
3. H1 = sha3_256( Public key)  => 32 byte
4. BitAddress = last 20 bytes of H1
5. Address = hx || HexString(BitAddress)
ex) hxaa688d74eb5f98b577883ca203535d2aa4f0838c

## Tested platform

We tested on window and macOS. If you find some problems on window, please send the report of the problems.



# Getting started

## Installation

The easiest way to install icxcli* is to use pip:

    $ pip install icxcli

or, if you are not installing in a virtualenv:

    $ sudo pip install icxcli

If you have the aws-cli installed and want to upgrade to the latest version you can run:

    $ pip install --upgrade icxcli

## Run CLI

Run command icli in the command line.  There are many sub commands like wallet create, wallet show, asset list, transfer for ICX service. You can get the help page by adding the command help.

    $ icli  --help
    usage:
            Normal commands:
                version
                help

            Wallet Commands:

                wallet create <file path> -p <password>
                wallet show <file path> -p <password>    | -n <network : mainnet | testnet | IP or domain>
                asset list <file path> -p <password>     | -n <network : mainnet | testnet | IP or domain>
                transfer <to> <amount> <file path> -p <password> -f <fee=10000000000000000> | -n  <network : mainnet | testnet | IP or domain>

            WARNING:

                Fee feature is the experimental feature; fee is fixed to 10000000000000000 loop for now so if you
                try to make a transaction with the modified fee, which is not 10000000000000000 loop, then you would
                not be able to make the transaction. you will be notified
                when it is possible to make a transaction with the modified fee.

            IF YOU MISS -n, icli WILL USE TESTNET.


    positional arguments:
      command           wallet create, wallet show, asset list, transfer

    optional arguments:
      -h, --help		show this help message and exit
      -p PASSWORD		password
      -f FEE		transaction fee
      -n NETWORK		mainnet or testnet or other IP or domain


## Console instructions

<table>
  <tr>
    <td>Command</td>
    <td>Function description</td>
  </tr>
  <tr>
    <td>version</td>
    <td>Shows the current software version. </td>
  </tr>
  <tr>
    <td>help</td>
    <td>Shows Help menu</td>
  </tr>
</table>



## Wallet operation

<table>
  <tr>
    <td>Command</td>
    <td>Function description</td>
  </tr>
  <tr>
    <td>wallet create &lt;file path&gt; -p &lt;password&gt;</td>
    <td>Create a wallet file.</td>
  </tr>
  <tr>
    <td>wallet show &lt;file path&gt; -p &lt;password&gt; -n &lt;network&gt;
</td>
    <td>Show current wallet information.</td>
  </tr>
  <tr>
    <td>asset list &lt;file path&gt; -p &lt;password&gt; -n &lt;network&gt;</td>
    <td>Enumerate the asset in the wallet. (ICX, ICX token) </td>
  </tr>
  <tr>
    <td>transfer &lt;to&gt; &lt;amount&gt; &lt;file path&gt; -p &lt;password&gt; -f &lt;fee&gt; -n &lt;network&gt;
</td>
    <td>Transfer the value to the specific address with the fee. </td>
  </tr>
</table>

### WARNING

Fee feature is the experimental feature; **fee is fixed to 10000000000000000 loop for now** so if you try to make a transaction with the modified fee, which is not 10000000000000000 loop, then you would not be able to make the transaction. you will be notified when it is possible to make a transaction with the modified fee.



## Create wallet file

    $ icli wallet create <file path> -p <password>

Create a wallet file with given wallet name, password and file path.

### Arguments

- file path : File path for the keystore file of the wallet.
- password: Password including alphabet character, number, and  special character. If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.

### Output

#### Successful case

Return 0 : Succeed to generate the keystore file for the wallet.

#### Error cases

**icli** will return following error code and message.

- Return 121: The file path is without a file name.
- Return 122: The file path is wrong.
- Return 123: The password is wrong.
- Return 124: The keystore file has already existed.
- Return 136: User doesn't have a permission to write the file.



## Show wallet information

    $ icli wallet show <file path> -p <password> -n <network>

Show wallet information.

### Arguments

- file path: File path for the keystore file of the wallet.
- password: Password including alphabet character, number, and  special character. If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
- network: 'testnet' or 'mainnet' or other IP or domain address.

### Output

Shows the all information of wallet.

- Wallet address
- Current balance
- Keystore file contents

#### Successful case

Return 0 : Print out wallet information including asset list.

#### Error cases

- Return 122: The file path is wrong.
- Return 123: The password is wrong.
- Return 126: The network is invalid.
- Return 133: The file is not a key store file.



## List up all assets in current wallet

    $ icli asset list <file path> -p <password> -n <network>

 Enumerate the list of all the assets of the wallet.

### Arguments

- file path : File path for the keystore file of the wallet.
- password: Password including alphabet character, number, and  special character. If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
- network: 'testnet' or 'mainnet' or other IP or domain address.

### Output

- List of all assets in current wallet.

#### Successful case

- Return 0 : Succeed to display.

#### Error cases

- Return 122: The file path is wrong.
- Return 123: The password is wrong.
- Return 126: The network is invalid.
- Return 133: The file is not a key store file.



## Transfer the value to the specific address with the fee.

    $ icli transfer <to> <amount> <file path> -p <password> -f <fee> -n <network>

Transfer the value from  A address to B address with the fee.

### Arguments

- to: Address of wallet to receive the asset.
- file path : File path for the keystore file of the wallet.
- password:  Password including alphabet character, number, and  special character. If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
- amount : Amount of money. (Unit:loop)
- fee : Transfer fee (Unit:loop)
    - **YOU SHOULD CHANGE BOTH THE UNIT OF AMOUNT AND FEE TO LOOP, ENTERING INTEGER VALUES FOR AMOUNT AND FEE NOT FLOATING NUMBERS.**
        - Unit: loop
            - Ex) 1 icx = 10<sup>18</sup> loop
- network: 'testnet' or 'mainnet' or other IP or domain address.

### Output

#### Successful case

Return 0 : Succeed to transfer

#### Error cases

**icli** will return following error code and message.

- Return 122: The file path is wrong.
- Return 123: The password is wrong.
- Return 126: The network is invalid.
- Return 127: The wallet doesn't have enough balance.
- Return 128: The fee is invalid.
- Return 130: The wallet address is wrong.
- Return 133: The file is not a key store file.
- Return 135: The amount is not integer values.




CHANGELOG.md
# Changelog


## [0.0.4] - Sep 10, 2018
### Added
* Support another domain or IP more than Mainnet and Testnet on ICLI.

### Changed
N/A

### Removed
N/A


## [0.0.3] - June 26, 2018
### Added
* Added the error when amount and fee user put are not integer values. 

### Changed
N/A

### Removed
N/A


## [0.0.2] - March 28, 2018
### Added

### Changed
* Use ```loop```, 10<sup>18</sup> icx , for value in ```transfer``` command, not ICX. 
* Removed ```-d``` in ```transfer``` command. 

### Removed
N/A


## [0.0.1] - March 13, 2018
### Added
 - Initialize version.

### Changed
N/A

### Removed
N/A


icxcli/cmd/wallet.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2018 ICON Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from enum import Enum
from icxcli.icx import wallet, NoEnoughBalanceInWallet, AddressIsWrong, PasswordIsWrong, FilePathIsWrong, \
    AmountIsInvalid, TransferFeeIsInvalid, FeeIsBiggerThanAmount, NotAKeyStoreFile, AddressIsSame, AmountIsNotInteger
import json


class ExitCode(Enum):
    """Exit codes for command line interface
    """
    SUCCEED = 0
    FILE_PATH_WITHOUT_FILE_NAME = 121
    FILE_PATH_IS_WRONG = 122
    PASSWORD_IS_WRONG = 123
    FILE_EXISTS = 124
    NETWORK_ID_IS_WRONG = 125
    NETWORK_IS_INVALID = 126
    WALLET_DOES_NOT_HAVE_ENOUGH_BALANCE = 127
    TRANSFER_FEE_IS_INVALID = 128
    TIMESTAMP_IS_NOT_CORRECT = 129
    WALLET_ADDRESS_IS_WRONG = 130
    AMOUNT_IS_INVALID = 131
    DECIMAL_POINT_INVALID = 132
    NOT_A_KEY_STORE_FILE = 133
    WALLET_ADDRESS_IS_SAME = 134
    AMOUNT_IS_NOT_INTEGER = 135
    NO_PERMISSION_TO_WRITE_FILE = 136


def create_wallet(password, file_path) -> int:
    """ Create a wallet file with given password and file path.

    :param password:  Password including alphabet character, number, and special character.
    If the user does not give password with -p, then CLI will show the prompt and user need to type the password.
    type(str)
    :param file_path: File path for the keystore file of the wallet. type(str)
    :return: Predefined exit code
    """
    try:
        wallet_info = wallet.create_wallet(password, file_path)
        print(f"Succeed to create wallet in {file_path}. ")
        print(f"Wallet address : {wallet_info.address} ")
        return ExitCode.SUCCEED.value
    except wallet.PasswordIsNotAcceptable:
        print("Fail: Password is not acceptable. ")
        print("Passwords must be at least 8 characters long including alphabet, number, and special character.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except wallet.FilePathIsWrong:
        print(f"Fail: Fail to open {file_path}. Change file path.")
        return ExitCode.FILE_PATH_IS_WRONG.value
    except wallet.NoPermissionToWriteFile:
        print(f"Fail: No permission to write {file_path}. Change file path.")
        return ExitCode.NO_PERMISSION_TO_WRITE_FILE.value
    except wallet.FileExists:
        print(f"Fail: {file_path} exists. Change keystore file name.")
        return ExitCode.FILE_EXISTS.value
    except wallet.FilePathWithoutFileName:
        print(f"Fail: File path needs a file name. Change file path.")
        return ExitCode.FILE_PATH_WITHOUT_FILE_NAME.value


def show_wallet(password, file_path, url) -> int:
    """ Show the all information of wallet. Show the balance and the information in keystore file.

    :param password:  Password including alphabet character, number, and special character. type(str)
    If the user does not give password with -p, then CLI will show the prompt and user need to type the password.
    :param file_path: File path for the keystore file of the wallet. type(str)
    :param url: api url. type(str)
    :return: Predefined exit code
    """
    try:
        wallet_address, balance, wallet_info = wallet.show_wallet(password, file_path, url)
        print(f"Succeed to show wallet in {file_path}. ")
        print(f"Wallet address : {wallet_address} ")
        print(f"Wallet balance : {balance} loop")
        wallet_info = json.dumps(wallet_info, indent=4, sort_keys=True)
        print(f"Wallet keystore_file_info : \n{wallet_info} ")
        return ExitCode.SUCCEED.value
    except wallet.NetworkIsInvalid:
        print(f"Fail: Network is invalid. It is impossible to connect {url}.")
        return ExitCode.NETWORK_IS_INVALID.value
    except wallet.PasswordIsNotAcceptable:
        print(f"Fail: Password is not acceptable. ")
        print(f"Password including alphabet character, number, and special character.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except wallet.FilePathIsWrong:
        print(f"Fail: Fail to open {file_path}. Change file path.")
        return ExitCode.FILE_PATH_IS_WRONG.value
    except PasswordIsWrong:
        print(f"Fail: Password is wrong.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except NotAKeyStoreFile:
        print(f"Fail: {file_path} is not a Key store File.")
        return ExitCode.NOT_A_KEY_STORE_FILE.value


def show_asset_list(password, file_path, url) -> int:
    """ Enumerate the list of all the assets of the wallet. Show the balance.
    :param password: Password including alphabet character, number, and special character.
    If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
    type(str)
    :param file_path: File path for the keystore file of the wallet. type(str)
    :param url: api url. type(str)
    :return: Predefined exit code
    """
    try:
        wallet_address, balance = wallet.show_asset_list(password, file_path, url)
        print(f"Succeed to show asset list in {file_path}. ")
        print(f"Wallet address : {wallet_address} ")
        print(f"Wallet balance : {balance} loop")
        return ExitCode.SUCCEED.value
    except wallet.NetworkIsInvalid:
        print(f"Fail: Network is invalid. It is impossible to connect {url}.")
        return ExitCode.NETWORK_IS_INVALID.value
    except wallet.PasswordIsNotAcceptable:
        print("Fail: Password is not acceptable. ")
        print("Password including alphabet character, number, and special character.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except wallet.FilePathIsWrong:
        print(f"Fail: Fail to open {file_path}. Change file path.")
        return ExitCode.FILE_PATH_IS_WRONG.value
    except PasswordIsWrong:
        print(f"Fail: Password is wrong.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except NotAKeyStoreFile:
        print(f"Fail: {file_path} is not a Key store File.")
        return ExitCode.NOT_A_KEY_STORE_FILE.value


def transfer_value_with_the_fee(password, fee, to, amount, file_path, url) -> int:
    """ Transfer the value to the specific address with the fee.

    :param password: Password including alphabet character, number, and special character.
    If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
    type(str)
    :param fee: Transaction fee.
    :param to: A user can change the decimal point to express all numbers including fee and amount.
    :param amount: A user can change the decimal point to express all numbers including fee and amount.
    :param file_path: File path for the keystore file of the wallet. type(str)
    :return: Predefined exit code
    """
    try:
        transfer_result = wallet.transfer_value_with_the_fee(password, fee, to, amount, file_path, url)
        print("Transaction has been completed successfully.")
        return ExitCode.SUCCEED.value
    except wallet.NetworkIsInvalid:
        print(f"Fail: Network is invalid. It is impossible to connect {url}.")
        return ExitCode.NETWORK_IS_INVALID.value
    except FilePathIsWrong:
        return ExitCode.FILE_PATH_IS_WRONG.value
    except AddressIsWrong:
        print(f"Fail: The transaction target address does not have the correct format. please check the address again.")
        return ExitCode.WALLET_ADDRESS_IS_WRONG.value
    except PasswordIsWrong:
        print(f"Fail: Password is wrong.")
        return ExitCode.PASSWORD_IS_WRONG.value
    except NoEnoughBalanceInWallet:
        print(f"Fail: Wallet does not have enough balance.")
        return ExitCode.WALLET_DOES_NOT_HAVE_ENOUGH_BALANCE.value
    except AmountIsInvalid:
        print(f"Fail: The amount you want to transfer is not valid.")
        return ExitCode.AMOUNT_IS_INVALID.value
    except TransferFeeIsInvalid:
        print(f"Fail: Transaction Fee is invalid. The fee should be 10000000000000000.")
        return ExitCode.TRANSFER_FEE_IS_INVALID.value
    except FeeIsBiggerThanAmount:
        print(f"Fail: Fee is bigger than transaction amount. Please check your fee again.")
        return ExitCode.TRANSFER_FEE_IS_INVALID.value
    except NotAKeyStoreFile:
        print(f"Fail: {file_path} is not a Key store File.")
        return ExitCode.NOT_A_KEY_STORE_FILE.value
    except AddressIsSame:
        print(f"Fail: Wallet address to transfer must be different from Wallet address to deposit.")
        return ExitCode.WALLET_ADDRESS_IS_SAME.value
    except AmountIsNotInteger:
        print(f"Fail: Amount should be integer and loop unit, not icx")
        return ExitCode.AMOUNT_IS_NOT_INTEGER.value


def store_wallet(file_path, json_string) -> int:
    """ Store wallet information file in JSON format.
    :param file_path: The path where the file will be saved. type(str)
    :param json_string: Contents of key_store_file. type(str)
    :return: Predefined exit code
    """
    pass


def make_key_store_content(password) -> int:
    """Make a content of key_store.

    :param password: Password including alphabet character, number, and special character.
    If the user doesn’t give password with -p, then CLI will show the prompt and user need to type the password.
    type(str)
    :return:
    key_store_content(dict)
    """
    pass
